### PR TITLE
[Create Block] Allows custom keys to be generated in block.json files and package.json files

### LINKED ISSUE #44626: Create Block: Allow templates to pass custom `block.json` keys
## What problem does this address?

It is currently difficult to create a custom Create Block Template for ACF Pro 6.0, because ACF requires there to be a custom key in the `block.json`, e.g. 

```
{
    "name": "acf-example-block",
    "title": "ACF Example Block",
    "description": "ACF example block",
    "apiVersion": 2,
    "acf": {
        "mode": "preview"
    }
}
```

## What is your proposed solution?

Allow the default values passed through to have an optional custom key. I'm not sure the extent of the updates that would needed to be done for Create Block, but imagine the entry point would be somewhere around here: https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/templates.js#L44

Ultimately, I would love to be able to write a custom Create Block Template with something like this:

```
const { join } = require( "path" );

/**
 * @link https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/init-block.js#L14
 */
module.exports = {
	pluginTemplatesPath: join(__dirname, 'plugin-templates'),
	blockTemplatesPath: join( __dirname, 'block-templates' ),
	defaultValues: {
		namespace: 'acf-blocks-starter',
		version: '0.1.0',
		category: 'acf-blocks-starter',
		editorStyle: 'file:./index.css',
		editorScript: 'file:./index.js',
		style: 'file:./style-index.css'
	},
	variants: {
		static: {},
		dynamic: {
			_acf: {
				mode: 'preview'
			}
		}
	}
};
```

Note the `_acf` key. I'm prefixing because I'm not sure what type of nomenclature or nesting we would want to help Create Block Templates core parse through custom keys and objects and `_` was the best I could come up with. I'm open to ideas and implementation. 

Then a user could scaffold out a custom ACF block with a custom Create Block Template ACF package and run something like: `npx @wordpress/create-block --template acf --variant dynamic --no-plugin --no-wp-scripts --category acf`

### PR DESCRIPTION
**Pull request by ryanwelcher**
## What?
This PR allows template authors to pass custom block.json properties by adding them to the `default` property or any `variant` and introduces a new key called `customPackageJSON` that will add keys to the generated package.json file.

Closes: #44626

## Why?
Block creation via the create-block package has limited options that can be passed via a custom template. With the functionality introduced by this PR, it becomes more flexible both when new features are added to other packages ( see https://github.com/WordPress/gutenberg/pull/44185 ) and especially for block developers who need/want to customize their process.

## Testing Instructions
This can be tested with a custom template or by changing the built-in template information in the package.

### Custom block.json keys
1. Add a the key/value pair to the default values in the [built-in templates](https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/templates.js#L44)
2. Run the command from the root of the package: `node ./index.js test-block`
3. Ensure that the key appears in the `block.json` file
4. Add a custom key to either of the [variations](https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/templates.js#L53)
5. Run the command from the root of the package: `node ./index.js test-block --variant=static|dynamic` ( use correct variant based on what you updated
6. Ensure that the key appears in the `block.json` file

### Custom package.json keys
1. Add `customPackageJSON` key to the [defaults](https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/templates.js#L44) with any  key/value pair
2. Run the command from the root of the package: `node ./index.js test-block`
3. Ensure that the resulting package.json contains the key/value pair

### CODE REVIEW

**Comment by colorful-tones** [comment]
@ryanwelcher this is awesomesauce!

I took it for a test spin and noticed an oddity. Here are my steps for testing (a bit different from yours):

1. `gh pr checkout 44649` (this branch)
2. `nvm use && npm ci`

I'm testing against my own [custom block template (download `.zip`)](https://github.com/colorful-tones/acf-create-block-template/archive/refs/heads/main.zip).

1. Download `.zip` of that block template and place (un-zipped) in a directory anywhere on the local hard drive, e.g. `~/new-test-plugin/block-template`
2. From `~/new-test-plugin` run `npm install --save ../path-to-gutenberg-repo-root`
3. Run `npx wp-create-block --template=./block-template/index.js --no-plugin --no-wp-scripts --variant=dynamic` from `~/new-test-plugin`

Once the block is output. I noticed that the `"variant": "dynamic"` is output, which seems odd. Here is the full output of my `block.json` (note the last line):

```
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 2,
	"name": "wpe/accordion",
	"version": "0.1.0",
	"title": "Accordion",
	"category": "text",
	"description": "An accordion",
	"textdomain": "wpe",
	"editorScript": "file:./index.js",
	"editorStyle": "file:./index.css",
	"style": "file:./style-index.css",
	"acf": {
		"mode": "preview"
	},
	"render": "file:./render.php",
	"variant": "dynamic"
}
```

Perhaps my [custom block template](https://github.com/colorful-tones/acf-create-block-template) is not written properly? Or I'm missing something?

**Comment by gziolo** [comment]
> Once the block is output. I noticed that the `"variant": "dynamic"` is output, which seems odd. Here is the full output of my `block.json` (note the last line):

I think it’s all good with your template, @colorful-tones. It looks like a side-effect of using the spread operator for fields passed from `defaultValues` to `block.json`. It’s probably safer to replicate the logic used for `package.json` that uses an explicit config value:

> Add `customPackageJSON` key to the [defaults](https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/lib/templates.js#L44) with any key/value pair

How do you both feel about offering `customBlockJSON` or `customBlockMetadata`? It might be more straightforward to document this way, too.

By the way, excellent job @ryanwelcher, taking care of the PR so fast!

**Comment by colorful-tones** [comment]
Makes sense about the spread operator.

I would vote for `customBlockJSON` and agree that having an explicit key assignment would be good overall.

**Comment by lgladdy** [comment]
Super excited about this PR. Agreed on `customBlockJSON` too. 🥳

**Comment by ryanwelcher** [comment]
Thanks for the feedback! I felt that the spread approach was a little naive so I agree that a dedicated key is probably the best approach. I'll add the `customBlockJSON` today.

**Comment by lgladdy** [comment]
@ryanwelcher Admittedly, this might need to be another ticket - but what would you think about adding support for `--template=@advancedcustomfields/<repo-name>` for example, so we could distribute a block template as a repo without folks needing to download files to use it?

**Comment by ryanwelcher** [comment]
> @ryanwelcher Admittedly, this might need to be another ticket - but what would you think about adding support for `--template=@advancedcustomfields/<repo-name>` for example, so we could distribute a block template as a repo without folks needing to download files to use it?

Interesting idea! Templates hosted on npm are currently supported by create-block. Does that suit the need or were you hoping to be able to use a git repo for example?

**Comment by lgladdy** [comment]
> > @ryanwelcher Admittedly, this might need to be another ticket - but what would you think about adding support for `--template=@advancedcustomfields/<repo-name>` for example, so we could distribute a block template as a repo without folks needing to download files to use it?
> 
> Interesting idea! Templates hosted on npm are currently supported by create-block. Does that suit the need or were you hoping to be able to use a git repo for example?

Ah, we can likely use the npm approach! I didn't realise that was even possible (and thought we'd need to get our users to download templates manually first etc). I'll check that out and open a new issue/PR if needed!

**Comment by ryanwelcher** [comment]
> Ah, we can likely use the npm approach! I didn't realise that was even possible (and thought we'd need to get our users to download templates manually first etc). I'll check that out and open a new issue/PR if needed!

Here is an [older example I made](https://www.npmjs.com/package/@ryanwelcher/dynamic-block-template) and the usage looks like:

`npx @wordpress/create-block example-plugin --template @ryanwelcher/dynamic-block-template`

Feel free to reach out with any questions!

**Comment by gziolo** [comment]
> @ryanwelcher Admittedly, this might need to be another ticket - but what would you think about adding support for `--template=@advancedcustomfields/<repo-name>` for example, so we could distribute a block template as a repo without folks needing to download files to use it?

@lgladdy, it might already work with GitHub URLs as it's supported by `npm install` out of the box:

https://docs.npmjs.com/cli/v8/commands/npm-install

Example:
`npm install github:<githubname>/<githubrepo>[#<commit-ish>]`

This is what we run behind the scenes:

https://github.com/WordPress/gutenberg/blob/154c0ee0516e84a9dcd17c144a7ffc7e8bbcb905/packages/create-block/lib/templates.js#L190-L192

**Review by colorful-tones — APPROVED** [review]
@ryanwelcher I'm not sure of the use case for `customPackageJSON`, _but_ it could potentially be useful and I'm just blanking on an immediate use case.

I tested and the `customBlockJSON` works wonderfully. Thanks for the great work.

**Review by gziolo — APPROVED** [review]
Great one 👍 We still should document new options in the docs folder so it’s easier to discover.

**Comment by ryanwelcher** [comment]
> @ryanwelcher I'm not sure of the use case for `customPackageJSON`, _but_ it could potentially be useful and I'm just blanking on an immediate use case.

One example is adding prettier config: `"prettier": "@wordpress/prettier-config"`, I'm sure there are others. This just allows customization which is never a bad thing 😄

File: packages/create-block/docs/external-template.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ The following configurable variables are used with the template files. Template
 -   `customScripts` (default: {}) – the list of custom scripts to add to `package.json` . It also allows overriding default scripts.
 -   `npmDependencies` (default: `[]`) – the list of remote npm packages to be installed in the project with [`npm install`](https://docs.npmjs.com/cli/v8/commands/npm-install) when `wpScripts` is enabled.
 -   `npmDevDependencies` (default: `[]`) – the list of remote npm packages to be installed in the project with [`npm install --save-dev`](https://docs.npmjs.com/cli/v8/commands/npm-install) when `wpScripts` is enabled.
+-   `customPackageJSON` (no default) - allows definition of additional properties for the generated package.json file.
```
Comment (gziolo) [inline packages/create-block/docs/external-template.md:77]: We have a bit of overlap with `customScripts`, `npmDependencies`, and `npmDevDependencies`. It doesn’t have to be a bad thing as we would have to put special care to handle them in a similar way through `customPackageJSON`. I’m mostly noticing here that this new option might override all other settings.

Comment (ryanwelcher) [reply packages/create-block/docs/external-template.md:77]: That's a good point. I wonder if we just filter those keys out of `customPackageJSON` before adding it?

Comment (gziolo) [reply packages/create-block/docs/external-template.md:77]: It isn’t a blocker. We can look into it once it starts to confuse people.

**Comment by bph** [comment]
Although create-block isn't really part of WordPress 6.2, Gutenberg 14.4 is part of the Gutenberg plugins version that are said to arrive at WordPress 6.2, so it would be a great occasion to let developers know about this enhancement in a Dev Note for the Fieldguide of WP6.2

**Pull request merged**